### PR TITLE
Superjump

### PR DESCRIPTION
**Pull request by TheGreenBandit**
A

### CODE REVIEW

File: src/backend/looped/self/super_jump.cpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace big
+{
+	class super_jump_looped : looped_command
+	{
+		using looped_command::looped_command;
+
+		virtual void on_tick() override
+		{
+			MISC::SET_SUPER_JUMP_THIS_FRAME(self::ped);
```
Comment (Yimura) [inline src/backend/looped/self/super_jump.cpp]: I remember that this would cause metrics from being sent out.

Comment (TheGreenBandit) [reply src/backend/looped/self/super_jump.cpp:13]: ill use the task_jump native then, that should work

**Comment by TheGreenBandit** [comment]
holdon forgot to make it check if the jump key is pressed from switching natives, right now it would be going in a infinite loop

**Comment by FireFox101889** [comment]
https://github.com/YimMenu/YimMenu/issues/108

**Comment by TheGreenBandit** [comment]
superjump should be ready now

**Comment by TheGreenBandit** [comment]
there is also this invisible car i coded that i could add to this pr or another, really all i did for it was change self::ped to self::veh 

 namespace big
{
	class invisibility : looped_command
	{
		using looped_command::looped_command;

		virtual void on_tick() override
		{
			ENTITY::SET_ENTITY_VISIBLE(self::veh, false, 0);
			if (g.self.local_visibility)
				NETWORK::SET_ENTITY_LOCALLY_VISIBLE(self::veh);
		}

		virtual void on_disable() override
		{
			ENTITY::SET_ENTITY_VISIBLE(self::veh, true, 0);
		}
	};

	invisibility g_invisibility("inviscar", "Invisiblity", "Makes your car invisible", g.vehicle.carinvisibility);
	bool_command g_local_visibility("localviscar", "Visible Locally", "Makes your car visible to yourself, other players will still not be able to see it", g.vehicle.localcar_visibility);
}

**Comment by TheGreenBandit** [comment]
just tested in game definitely needs improvement

**Comment by TheGreenBandit** [comment]
it looks as though its actually doing the beast jump based off how the ped is animated

**Comment by TheGreenBandit** [comment]
Coded this from my phone

```#include "natives.hpp"
#include "backend/looped_command.hpp"

namespace big
{
	class super_jump_looped : looped_command
	{
		using looped_command::looped_command;

		virtual void on_tick() override
		{
                    if(PAD::IS_CONTROL_JUST_PRESSED(0, (int)ControllerInputs::INPUT_JUMP) && !ENTITY::IS_ENTITY_IN_AIR(self::ped) && !ENTITY::IS_ENTITY_IN_VEHICLE(self::ped))
                    {
                         ENTITY::APPLY_FORCE_TO_ENTITY_CENTER_OF_MASS(1, 0, 0, 1, true, true, true)
                    }
                    if(PAD::IS_CONTROL_JUST_PRESSED(0, (int)ControllerInputs::INPUT_VEH_ACCELERATE) && ENTITY::IS_ENTITY_IN_AIR(self::ped) && !ENTITY::IS_ENTITY_IN_VEHICLE(self::ped))
                    {
                         ENTITY::APPLY_FORCE_TO_ENTITY_CENTER_OF_MASS(0, 0, 5, 0, true, true, true) //5 may have to be changed to a higher value
                    }
		}
	};

	super_jump_looped g_super_jump_looped("superjump", "Super Jump", "Allows You To Jump Really High", g.self.super_jump);
}```

Has to be tested

**Comment by SkiddyToast** [comment]
I mean you could just add a hook to the `CTaskJump` constructor and set the flags there. Would bypass the need to use frame flags or use a native.

**Comment by TheGreenBandit** [comment]
True but using apply force there would be no way to have it detected, and if we want we can set custom input values for it

**Comment by SkiddyToast** [comment]
There would be no reason the `CTaskJump` hook would be detected, and you can use the `CTaskJump::Tunables` to edit everything about the jump.

**Comment by TheGreenBandit** [comment]
> There would be no reason the `CTaskJump` hook would be detected, and you can use [`CTaskJump::Tunables`](https://alexguirre.github.io/rage-parser-dumps/dump.html?game=gta5&build=2802#CTaskJump__Tunables) to edit everything about the jump.

Alright, i will code it with that later.

**Comment by Aure7138** [comment]
> > There would be no reason the `CTaskJump` hook would be detected, and you can use [`CTaskJump::Tunables`](https://alexguirre.github.io/rage-parser-dumps/dump.html?game=gta5&build=2802#CTaskJump__Tunables) to edit everything about the jump.
> 
> Alright, i will check it out
> 
> edit:: i dont see anything about calling the super jump nor have i any idea how i would use it

@SkiddyToast actually gives almost everything, I just give a boost
```cpp
48 89 5C 24 ? 89 54 24 10 57 48 83 EC 30 0F 29 74 24

std::uint64_t taskjump_constructor(std::uint64_t a1, int a2)
{
    if (super_jump)
        a2 |= 1 << 15;
    if (beast_jump)
        a2 |= (1 << 15) | (1 << 17);
    return m_taskjump_constructor->get_original<decltype(&taskjump_constructor)>()(a1, a2);
}
```
I'm currently looking for a way to check self ped, not sure if this is needed

**Comment by TheGreenBandit** [comment]
> > There would be no reason the `CTaskJump` hook would be detected, and you can use [`CTaskJump::Tunables`](https://alexguirre.github.io/rage-parser-dumps/dump.html?game=gta5&build=2802#CTaskJump__Tunables) to edit everything about the jump.
> 
> Alright, i will check it out
> 
> edit:: i dont see anything about calling the super jump nor have i any idea how i would use it

heres a fixed version of my apply force to entity, i removed one apply force cause its not really needed ```

```if (PAD::IS_CONTROL_JUST_PRESSED(0, (int)ControllerInputs::INPUT_JUMP) && !ENTITY::IS_ENTITY_IN_AIR(self::ped) && !PED::IS_PED_IN_ANY_VEHICLE(self::ped, NULL))
			{
				ENTITY::APPLY_FORCE_TO_ENTITY_CENTER_OF_MASS(self::ped, 0, 0, 10000, 5000, true, true, true, false);
			}```
p8 is is mass related, setting it to false allows going forward while in air letting the superjump go forward

only problem with this is even with ragdoll turned off you will still get the animation of getting knocked over

**Comment by TheGreenBandit** [comment]
> > > There would be no reason the `CTaskJump` hook would be detected, and you can use [`CTaskJump::Tunables`](https://alexguirre.github.io/rage-parser-dumps/dump.html?game=gta5&build=2802#CTaskJump__Tunables) to edit everything about the jump.
> > 
> > 
> > Alright, i will check it out
> > edit:: i dont see anything about calling the super jump nor have i any idea how i would use it
> 
> @SkiddyToast actually gives almost everything, I just give a boost
> 
> ```c++
> 48 89 5C 24 ? 89 54 24 10 57 48 83 EC 30 0F 29 74 24
> 
> std::uint64_t taskjump_constructor(std::uint64_t a1, int a2)
> {
>     if (super_jump)
>         a2 |= 1 << 15;
>     if (beast_jump)
>         a2 |= (1 << 15) | (1 << 17);
>     return m_taskjump_constructor->get_original<decltype(&taskjump_constructor)>()(a1, a2);
> }
> ```
> 
> I'm currently looking for a way to check self ped, not sure if this is needed

oh yeah i have no idea how to find the pointers for things like this

**Comment by Aure7138** [comment]
> oh yeah i have no idea how to find the pointers for things like this

ClassInformer and 1.63 or earlier dumps

**Comment by TheGreenBandit** [comment]
YimMenu - Microsoft Visual Studio 1_4_2023 5_11_23 PM
like this?

**Comment by Aure7138** [comment]
> (sorry for being a bit slow i only really know how to use natives)

Almost, there is a mistake in the picture

**Comment by TheGreenBandit** [comment]
> Almost, there is a mistake in the picture

is it the         return g_hooking->get_original<hooks::taskjump_constructor>()(a1, a2); ?

**Comment by Aure7138** [comment]
> > Almost, there is a mistake in the picture
> 
> is it the return g_hooking->get_originalhooks::taskjump_constructor()(a1, a2); ?

```cpp
std::uint64_t hooks::taskjump_constructor(std::uint64_t a1, int a2)
{
    stuff
}
```

**Comment by TheGreenBandit** [comment]
> std::uint64_t hooks::taskjump_constructor(std::uint64_t a1, int a2)
> {
>     stuff
> }
Wait so i would put the stuff from 
(Link)[https://alexguirre.github.io/rage-parser-dumps/dump.html?game=gta5&build=2802&search=jump#CTaskJump__Tunables] in there?

**Comment by Aure7138** [comment]
uh no that's another class

**Comment by TheGreenBandit** [comment]
> ```c++
>  return m_taskjump_constructor->get_original<decltype(&taskjump_constructor)>()(a1, a2);
> ```

i got no idea then

**Comment by FireFox101889** [comment]
The return type.

**Comment by TheGreenBandit** [comment]
ok put it as  			m_taskjump_constructor = ptr.as<functions::taskjump_constructor>(); in pointers now but what the hell do i do in function types

**Comment by Gh0sSte** [comment]
task jump should not be detected. It isnt a frame flag and from my testing mod menus cannot detect it

**Comment by TheGreenBandit** [comment]
> task jump should not be detected. It isnt a frame flag and from my testing mod menus cannot detect it

Yeah but I'm trying to get superjump, i have yet to push an updated commit to this that renames the current superjump to its appropriate name beast jump and creates a artifical superjump using applyforcetoentity. I'm working on debugging when the ped hits the ground how they fall over with no ragdoll on

**Comment by TheGreenBandit** [comment]
pushed a commit renameing the original task_jump way that i was gonna do to beast jump and use APPLY_FORCE_TO_ENTITY_CENTER_OF_MASS for artificial superjump, ped tasks clear every frame so the ped doesn't ragdoll when it hits ground

**Comment by FireFox101889** [comment]
> > ```c++
> >  return m_taskjump_constructor->get_original<decltype(&taskjump_constructor)>()(a1, a2);
> > ```
> 
> i got no idea then

Aure's sig and hook example works perfect for me, you're being spooned enough. please figure it out, i gave you the hint that it had to do with your return type. please refer to Aure's example again.

**Comment by TheGreenBandit** [comment]
> > > ```c++
> > >  return m_taskjump_constructor->get_original<decltype(&taskjump_constructor)>()(a1, a2);
> > > ```
> > 
> > 
> > i got no idea then
> 
> Aure's sig and hook example works perfect for me, you're being spooned enough. please figure it out, i gave you the hint that it had to do with your return type. please refer to Aure's example again.

i have no experience with it, I'd rather just do it to where i have it figured out and working than try to get something i know nothing about to work

**Comment by TheGreenBandit** [comment]
This is ready to go Yimura unless you have any comments

File: src/backend/looped/self/super_jump.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	{
+		using looped_command::looped_command;
+
+		virtual void on_tick() override
+		{
+			if (PAD::IS_CONTROL_JUST_PRESSED(0, (int)ControllerInputs::INPUT_JUMP) && !PED::IS_PED_IN_ANY_VEHICLE(self::ped, NULL) && !ENTITY::IS_ENTITY_IN_AIR(self::ped))
+			{
+				ENTITY::APPLY_FORCE_TO_ENTITY_CENTER_OF_MASS(self::ped, 0, 0, 10000, 5000, true, true, true, false);
+			}
+			if (PED::IS_PED_FALLING(self::ped))
```
Comment (MarkEcza) [inline src/backend/looped/self/super_jump.cpp:18]: If your worried about ragdolling you can use ``SET_PED_CAN_RAGDOLL`` & ``SET_PED_CAN_RAGDOLL_FROM_PLAYER_IMPACT``.

Comment (TheGreenBandit) [reply src/backend/looped/self/super_jump.cpp:18]: I will see if that has the same effect as my version when i get out of school. Ill check it out in less than 2 hours

Comment (TheGreenBandit) [reply src/backend/looped/self/super_jump.cpp:18]: nope, the reason i have clear tasks is so the following animation doesn't occur: 

https://user-images.githubusercontent.com/106003542/211103428-b3bcf3f0-6db6-49aa-b70a-b1d3719ea2d5.mp4

**Review by Yimura — APPROVED** [review]

**Pull request merged**